feat: Fixes #74. Allow upload/download JSON/private key from accounts

ltfschoen opened the pull request:
* Upload JSON/private keys. Decrypt in modal by entered password to unlock secret key and store in memory and local storage. Error notifications on password input field when empty or incorrect.
* Download JSON/private keys. Decrypt in modal by entered password to unlock secret key then download encrypted copy. Error notifications on password input field when empty or incorrect.
![screen shot 2018-08-07 at 3 33 09 am](https://user-images.githubusercontent.com/6226175/43749055-cd9688b0-99f2-11e8-97c0-b004f0cd67b2.png)
* Requires `pair.lock()` https://github.com/polkadot-js/apps/pull/198 functionality to be incorporated at the following FIXME locations in the code so that the secret key is locked again immediately after each individual download or upload, otherwise another user could repeat the process and download or upload with the incorrect password.
  * https://github.com/polkadot-js/apps/compare/luke-74-download-upload-with-password?expand=1#diff-fdef2d47f48cbd87a974f75935042996R37
  * https://github.com/polkadot-js/apps/compare/luke-74-download-upload-with-password?expand=1#diff-fdef2d47f48cbd87a974f75935042996R50
  * https://github.com/polkadot-js/apps/compare/luke-74-download-upload-with-password?expand=1#diff-4460fae0eb0b78fa6efa8b092a9a0991R32
  * https://github.com/polkadot-js/apps/compare/luke-74-download-upload-with-password?expand=1#diff-fcf70b0c98d01219eab7a943d0de67a6R52
* All tests pass and tslint errors resolved

jacogr commented on tsconfig.json:
```diff
@@ -23,7 +23,8 @@
       "./node_modules/@polkadot/ts",
-      "./node_modules/@types"
+      "./node_modules/@types",
+      "packages/ui-app/src/@types"
```
No. Only in root.

ltfschoen commented on packages/ui-app/src/DownloadButton.tsx:
```diff
@@ -0,0 +1,273 @@
+
+    const keyringAddress: KeyringAddress = keyring.getAddress(address);
+
+    let translateError = null;
```
So there would be two blank lines before the line with `let translateError = null;`?

ltfschoen commented:
@jacogr I've closed out the following issues:

* [x] - all translatable strings

* [x] - rename handleAccountChange to onAccountChange

* [x] - add license header to 
  * packages/ui-app/src/@types/file-saver/index.d.ts
  * packages/ui-app/src/@types/react-file-reader/index.d.ts

* [x] - fix anti-pattern where i was getting stuff that can fail inside render. This is a map function on the class

* [x] - remove unnecessary `return`'s'

* [x] - concat the ui--Password with className and pass it instead of wrapping

* [x] - remove `"packages/ui-app/src/@types"` type dependency from root tsconfig.json

* [x] - rename to just `${address}.json`

* [x] - remove `/// <reference path="./@types/file-saver/index.d.ts" />` as we do not do this. use @types types instead renaming both file-saver and react-file-reader files to types.d.ts

* [x] - Ant-pattern. focus() is done with ref

* [x] - Why are these duplicated? Rather pull-in the types from where already defined. (re: `size?: 'mini' | 'tiny' | 'small' | 'large' | 'big' | 'huge' | 'massive'`)

* [x] - Notification to be part of Input (optional), not separate and just used for Password. TODO is remove `// <Notification error={error} />` and move functionality elsewhere

* [x] - Initial run-through. It does not look specific to Account. It would be better served with the buttons specific and just below the display in app-account. add two buttons (upload/download) after AddressSummary component

* [x] - Upload is a account creation function (i.e. part of creator), not part of summary which gets shown everywhere. (re: `showUploadAndDownloadButtons(address, handleChangeAccount)` in AddressSummary.tsx).
Additionally, this is a shared address (i.e. not-account-specific) component, shared everywhere. So in which cases do we need upload/download for addresses? (i.e. on the staking screen). It is specific the the app-account not generic for all displays everywhere.

* [x] - fix translations missing from DownloadButton.tsx and UploadButton.tsx, and remove the FIXME refresh comment

* [x] - move `shortValue = `${address.slice(0, 7)}…${address.slice(-7)}`;` into common utility

* [x] - SUI does do a way of doing this. Additionally, this is not password specific, it is input specific.

* [x] - move new Upload and Download icons so not in AddressSummary component

ltfschoen commented:
@jacogr Outstanding tasks are as follows: I'm not sure how to best resolve the following TSLint error, and I'm not sure why if immediately after creating an account I try to Download the JSON file, it gives me the following error... any tips appreciated!. I also understand you'd like the two icons below the AccountSummary component?

* [ ] - find out why after create new account it doesn't accept the password when try to Download the JSON of the account just created, it gives error `ExtError: No encrypted data available to decode` from decode.ts, because it's expecting en `encoded` value as a second parameter??)

* [ ] - fix TSLint error `ERROR: /Users/Ls/code/blockchain/clones/paritytech/apps/packages/app-accounts/src/UploadButton.tsx[76, 19]: 'pair' is declared but its value is never read.`, since it won't let me not declare it and makes me use it when i declare it (but i don't need to)

* [ ] - use flexbox to move buttons below the AddressSummary component

jacogr commented:
Ok, that one is interesting (actually the re-lock account PR https://github.com/polkadot-js/apps/pull/198) will have the same issue. So let's think through what happens -

1. User opens browser, the `loadAll` in ui-keyring loads all accounts via `fromJson` - here the encoded value is available, but the seed/secretKey is not (at least not until unlocked)
2. User goes to accounts, creates a new one, in this case it is added via `addFromSeed` - so the seed/secretKey is available in memory, not the encoded value

The quick solution would be to load the account with the encoded value (ui-keyring/src/account/save.ts) just after the json is created. Basically, pass the encoded values that has just been encoded, back into the keyring.

jacogr commented on packages/app-accounts/src/UploadButton.tsx:
```diff
@@ -0,0 +1,324 @@
+      return null;
+    }
+
+    const keyringAddress: KeyringAddress = keyring.getAddress(uploadedFileKeyringPair.address);
```
just `const keyringAddress = `

jacogr commented on packages/app-accounts/src/UploadButton.tsx:
```diff
@@ -0,0 +1,324 @@
+  }
+
+  nextState (newState: State): void {
+    this.setState(
```
As per the other vomponent, is there a specific reason for `nextState`, does not look like any state mutation takes place. `setState` will do fine and remove extra noise.

jacogr commented on packages/app-accounts/src/UploadButton.tsx:
```diff
@@ -0,0 +1,324 @@
+  }
+
+  onChangePassword = (password: string): void => {
+    this.nextState({ password, unlockError: null } as State);
```
Copy-and-paste issue here - once again 2 x setState, combine into one only.

jacogr commented on packages/app-accounts/src/UploadButton.tsx:
```diff
@@ -0,0 +1,324 @@
+      return;
+    }
+
+    this.processUploadedFileStorage();
```
Does this need to be split? Called from different places?

jacogr commented on packages/app-accounts/src/index.css:
```diff
@@ -10,3 +10,30 @@
+  padding-top: 1.5em;
+}
+
+.flex-group-row {
```
Since these are not global, prefix with `.accounts--` to avoid collissions.

jacogr commented on packages/app-accounts/src/util/getAddressShortValue.tsx:
```diff
@@ -0,0 +1,13 @@
+
+import isUndefined from '@polkadot/util/is/undefined';
+
+export default function getAddressShortValue (address: string): string {
```
ui-app/util/toShortAddress

jacogr commented on packages/ui-app/src/@types/file-saver/types.d.ts:
```diff
@@ -0,0 +1,6 @@
+// of the ISC license. See the LICENSE file for details.
+
+// https://www.typescriptlang.org/docs/handbook/namespaces-and-modules.html
+declare module 'file-saver';
```
If these definitions are not available on definately-typed, they need to be added to the polkadot-js/ts repo.

jacogr commented on packages/ui-app/src/@types/react-file-reader/types.d.ts:
```diff
@@ -0,0 +1,5 @@
+// This software may be modified and distributed under the terms
+// of the ISC license. See the LICENSE file for details.
+
+declare module 'react-file-reader';
```
As previous comment for the types.

jacogr commented on packages/ui-app/src/Input.tsx:
```diff
@@ -2,20 +2,23 @@
 // of the ISC license. See the LICENSE file for details.
 
-import { BareProps } from './types';
+import { BareProps, UnlockI18n } from './types';
```
Here surely it cannot be `UnlockI18N`? That sounds specific to locking - rather is needs to ahve a proper name to indicate use.

jacogr replied:
Additionally, why is the string not rendered on the caller (via t(...) or anything else) and this prop just purely a string or React.ReactNode?

ltfschoen replied:
Thanks. I've created type called `InputErrorMessage` and a component called `InputError` that accepts that type as a prop

jacogr commented on packages/ui-app/src/Input.tsx:
```diff
@@ -49,48 +52,52 @@ export default class Input extends React.PureComponent<Props, State> {
-          placeholder={placeholder}
-          type={type}
-          value={value}
+      <div>
```
Extra div, should not have been added - remove.

jacogr commented on packages/ui-app/src/Notification.tsx:
```diff
@@ -0,0 +1,44 @@
+import React from 'react';
+
+import Message from 'semantic-ui-react/dist/commonjs/collections/Message/index';
+// import { SemanticSIZES } from 'semantic-ui-react/dist/commonjs/index';
```
Remove line, unused code. (If there is a specific reason, indicate it as such with an extra note)

jacogr commented on packages/ui-app/src/Notification.tsx:
```diff
@@ -0,0 +1,44 @@
+    const renderMessageBody = (error: UnlockI18n) => {
+      return (
+        <div>
+          {/* <Message.Header>{error.key}</Message.Header> */}
```
? Remove.

jacogr commented on packages/ui-app/src/Notification.tsx:
```diff
@@ -0,0 +1,44 @@
+      return (
+        <div>
+          {/* <Message.Header>{error.key}</Message.Header> */}
+          <p>{error.key}: <b>{error.value}</b></p>
```
I18N is not used like this at all. Hence earlier comments, remove this key/value, it should be a pre React.ReactNode

jacogr commented on packages/ui-app/src/Password.tsx:
```diff
@@ -2,18 +2,22 @@
 // of the ISC license. See the LICENSE file for details.
 
-import { BareProps } from './types';
+import { BareProps, UnlockI18n } from './types';
```
As per the comments before on the use of this.

jacogr commented on packages/ui-app/src/Password.tsx:
```diff
@@ -2,18 +2,22 @@
   children?: React.ReactNode,
   isDisabled?: boolean,
   isError?: boolean,
+  error?: UnlockI18n | null,
```
string or Recat.ReactNode (preferred)

jacogr commented on packages/ui-app/src/types.d.ts:
```diff
@@ -25,3 +25,8 @@ export type BaseContext = {
   }
 };
+
+export type UnlockI18n = {
```
Remove if we are not passing these through - as it should be React.ReactNode going in

ltfschoen replied:
Existing code prior to this PR uses `UnlockI18n` in ui-signer and app-toolbox, I've left that implementation unchanged... or should I change that to `React.ReactNode` too?

jacogr commented on packages/ui-keyring/src/account/backup.ts:
```diff
@@ -0,0 +1,39 @@
+    return jsonDecrypted;
+  } catch (error) {
+    console.error('Unable to decrypt account with given password: ', error);
+    return;
```
Unneeded return.

jacogr commented on packages/ui-signer/src/Unlock.tsx:
```diff
@@ -2,7 +2,7 @@
 // of the ISC license. See the LICENSE file for details.
 
-import { I18nProps } from '@polkadot/ui-app/types';
+import { I18nProps, UnlockI18n } from '@polkadot/ui-app/types';
```
As per above comments.

jacogr commented on packages/ui-signer/src/Unlock.tsx:
```diff
@@ -14,14 +14,17 @@ import keyring from '@polkadot/ui-keyring/index';
 type Props = I18nProps & {
-  error?: string,
+  autoFocus?: boolean,
+  error?: UnlockI18n | null,
```
Restore as it was.

ltfschoen replied:
I'll be passing through InputErrorMessage type now

jacogr commented on packages/ui-signer/src/Unlock.tsx:
```diff
@@ -14,14 +14,17 @@ import keyring from '@polkadot/ui-keyring/index';
+  error?: UnlockI18n | null,
   onChange: (password: string) => void,
   password: string,
+  passwordWidth?: string,
```
Classes/styles should take care of this. Pass in correct class rather.

ltfschoen commented on packages/app-accounts/src/DownloadButton.tsx:
```diff
@@ -0,0 +1,273 @@
+import keyring from '@polkadot/ui-keyring/index';
+import IdentityIcon from '@polkadot/ui-react/IdentityIcon';
+import isUndefined from '@polkadot/util/is/undefined';
+import getAddressShortValue from './util/getAddressShortValue';
```
i'm changing to:
```
...
import React from 'react';
import FileSaver from 'file-saver';
import keyring from '@polkadot/ui-keyring/index';
import IdentityIcon from '@polkadot/ui-react/IdentityIcon';
import isUndefined from '@polkadot/util/is/undefined';

import Button from '@polkadot/ui-app/Button';
import Modal from '@polkadot/ui-app/Modal';
import Unlock from '@polkadot/ui-signer/Unlock';
import getAddressShortValue from './util/getAddressShortValue';
...
```

ltfschoen commented:
@jacogr I've addressed all your review comments and pushed the latest changes after rebased with latest from master branch.

I'm working through the following now:

* [ ] - Instead of investing time into creating a type definition on DefinatelyTyped (which I started doing [here](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/27999) ), i'm instead going to proceed with your recommendation of replacing react-file-reader with react-dropzone, which we're already using.

* [ ] - Adding unit tests

jacogr commented on packages/app-accounts/src/DownloadButton.spec.js:
```diff
@@ -0,0 +1,17 @@
+import React from 'react';
```
Header?

jacogr replied:
Header?

jacogr commented on packages/app-accounts/src/DownloadButton.tsx:
```diff
@@ -0,0 +1,213 @@
+
+            this.hidePasswordModal();
+          } else {
+            this.setState({ error: { key: t('error'), value: t('Unable to obtain account from memory') } });
```
setState inside setState - something is not right. Do the checks first and then a single setState.

jacogr commented on packages/app-accounts/src/DownloadButton.tsx:
```diff
@@ -0,0 +1,213 @@
+
+    return (
+      <div className={'accounts--Address-download'}>
+        { isPasswordModalOpen ? (
```
The Modal also has `isPasswordModalOpen` - suggest using that only?

jacogr commented on packages/app-accounts/src/DownloadButton.tsx:
```diff
@@ -0,0 +1,213 @@
+              <Modal.Content>
+                <div className='ui--grid'>
+                  <div className={'accounts--Address-modal'}>
+                    <AddressMini isShort={true} value={address} />
```
Just `isShort` will do the trick here.

Is this not too small? Row or Summary maybe? (We can hide the balances/transactions if need-be))

ltfschoen replied:
AddressMini is quite small, would be good for small screen sizes, i'll see how the other's look

ltfschoen replied:
I just tried implementing both AddressSummary and AddressMini in both DownloadButton.tsx and UploadButton.tsx. I was going to use media queries to toggle one or the other depending on the device width in use, or even try something like the react-responsive library.

I implemented it as follows:
```
import AddressMini from '@polkadot/ui-app/AddressMini';
import AddressSummary from '@polkadot/ui-app/AddressSummary';
...
              <div className={'accounts--Address-modal'}>
                <AddressMini isShort value={address} />
                <AddressSummary value={address} />
```

But when I added the AddressSummary component it gave error `Uncaught ExtError: Application root must be wrapped inside 'rx-react/Api' to provide API context`.

So I added the following, thinking it would resolve the error: 
```
import withApi from '@polkadot/ui-react-rx/with/api';

...

const Component: React.ComponentType<any> = translate(
  withApi(DownloadButton)
);

export default Component;
```

But it didn't resolve the error. What do you think?

jacogr commented on packages/app-accounts/src/DownloadButton.tsx:
```diff
@@ -0,0 +1,213 @@
+    return (
+      <Unlock
+        autoFocus={true}
+        error={translateError || undefined}
```
if it is not set, it is `undefined` already/

jacogr commented on packages/app-accounts/src/DownloadButton.tsx:
```diff
@@ -0,0 +1,213 @@
+    }
+
+    const keyringAddress = keyring.getAddress(address);
+    let translateError: InputErrorMessage | undefined = undefined;
```
No need to `= undefined`, that is what it is.

jacogr commented on packages/app-accounts/src/DownloadButton.tsx:
```diff
@@ -0,0 +1,213 @@
+          />
+          <Button.Or />
+          <Button
+            isDisabled={false}
```
No need to set this.

jacogr commented on packages/app-accounts/src/Editor.tsx:
```diff
@@ -98,21 +100,36 @@ class Editor extends React.PureComponent<Props, State> {
-      });
+      return (
+        <div>
+          <div>{t('There are no saved accounts. Create an account or upload a JSON file of a saved account.')}</div>
```
No. As elsewhere, construct is -

`t(<key>, { defaultValue: <actual text> })`

jacogr commented on packages/app-accounts/src/Editor.tsx:
```diff
@@ -98,21 +100,36 @@ class Editor extends React.PureComponent<Props, State> {
+          <div>{t('There are no saved accounts. Create an account or upload a JSON file of a saved account.')}</div>
+          <div className={'accounts--Address-wrapper'}>
+            <div className={'accounts--Address-file'}>
+              <UploadButton size={'big'} onChangeAccount={this.onChangeAccount} />
```
props on seperate lines. `size='big'`

jacogr commented on packages/app-accounts/src/Editor.tsx:
```diff
@@ -98,21 +100,36 @@ class Editor extends React.PureComponent<Props, State> {
-        />
-        <div className='grow'>
-          <div className='ui--row'>
+      <div className={'accounts--flex-group-row'}>
```
No. `className='accounts--flex-grow'`

ltfschoen replied:
Sorry I don't quite understand. Are you suggesting I just rename that className to `accounts--flex-grow`, or would you like me to somehow incorporate `flex-grow` https://developer.mozilla.org/en-US/docs/Web/CSS/flex-grow?

jacogr replied:
Now I get the comment.

`<div className='accounts--flex-group-row'>` (without the wrapping {}'s)

jacogr commented on packages/app-accounts/src/Editor.tsx:
```diff
@@ -98,21 +100,36 @@ class Editor extends React.PureComponent<Props, State> {
+          </div>
+        </div>
+        <div className={'accounts--flex-container-col-inputs'}>
+          <div className={'accounts--flex-item'}>
```
Fix all the classes from this point to the previous comment, moving up.

jacogr commented on packages/app-accounts/src/Editor.tsx:
```diff
@@ -126,7 +143,7 @@ class Editor extends React.PureComponent<Props, State> {
             />
           </div>
-          <div className='ui--row'>
+          <div className={'accounts--flex-item'}>
```
As above.

jacogr replied:
No need for wrapping {}'s

jacogr commented on packages/app-accounts/src/UploadButton.tsx:
```diff
@@ -0,0 +1,259 @@
+
+          return;
+        } else {
+          this.setState({ error: { key: t('error'), value: t('Unable to upload account into memory') } });
```
setState inside setState callback, this can't be right.

Seem to be making _exactly_ the same comments in 2 files here - DRY it up and share as needed, this seems like a lot of copy-and-pasting going on.

jacogr commented on packages/app-accounts/src/UploadButton.tsx:
```diff
@@ -0,0 +1,259 @@
+
+    return (
+      <div className={'accounts--Address-upload'}>
+        { address && isPasswordModalOpen ? (
```
As per the comment in the other file on the boolean flag.

jacogr commented on packages/app-accounts/src/UploadButton.tsx:
```diff
@@ -0,0 +1,259 @@
+              dimmer='inverted'
+              open={isPasswordModalOpen}
+              onClose={this.hidePasswordModal}
+              size={'mini'}
```
No, `size='mini'`

jacogr commented on packages/app-accounts/src/UploadButton.tsx:
```diff
@@ -0,0 +1,259 @@
+            >
+              <Modal.Content>
+                <div className='ui--grid'>
+                  <div className={'accounts--Address-modal'}>
```
Same `className=` comments as per the previous file. Change form this line downwards ...

jacogr commented on packages/app-accounts/src/UploadButton.tsx:
```diff
@@ -0,0 +1,259 @@
+  renderContent () {
+    const { error, password, uploadedFileKeyringPair } = this.state;
+    const { t } = this.props;
+    let translateError: InputErrorMessage | undefined = undefined;
```
`let translateError: InputErrorMessage;` is the same and preferrable, no need to `= undefined`

jacogr commented on packages/ui-app/src/types.d.ts:
```diff
@@ -25,3 +25,13 @@ export type BaseContext = {
+  value: any
+};
+
+export type InputErrorMessage = {
```
Ok. I did comment on this in the previous iteration - the error needs to be defined as `React.ReactNode` as per the previous comment. NOT this.

Can we please get it right. The translation happens on the calling component and calling component only.

jacogr commented on packages/ui-app/src/util/toShortAddress.ts:
```diff
@@ -2,6 +2,8 @@
+
 export default function toShortAddress (address: string): string {
-  return `${address.slice(0, 7)}…${address.slice(-7)}`;
+  return isUndefined(address) ? '' : `${address.slice(0, 7)}…${address.slice(-7)}`;
```
Use toShortAddress as per the previous review comments.

ltfschoen replied:
Yes I have, this is the `toShortAddress` function that I'm using now

jacogr commented on packages/ui-keyring/src/account/backup.ts:
```diff
@@ -0,0 +1,38 @@
+    return;
+  }
+
+  const pair: KeyringPair = keyring.getPair(_address);
```
`const pair = ` - no need to add the type, TS gets it. (Or should)

jacogr commented on packages/ui-keyring/src/account/restore.ts:
```diff
@@ -0,0 +1,40 @@
+  try {
+    pair.decodePkcs8(password);
+
+    store.set(accountKey(_address), json);
```
Extra line before.

ltfschoen replied:
So I should get rid of Line 30? i.e.
```
  try {
    pair.decodePkcs8(password);
    store.set(accountKey(_address), json);
...
```

jacogr commented on packages/ui-keyring/src/index.ts:
```diff
@@ -34,6 +36,10 @@ const state: State = {
 
 export default ({
+  addFromJson: (json: KeyringPair$Json): KeyringPair =>
+    state.keyring.addFromAddress(json.address, json.meta),
```
Why in this form?

jacogr commented on packages/ui-signer/src/Unlock.tsx:
```diff
@@ -14,14 +14,16 @@ import keyring from '@polkadot/ui-keyring/index';
 type Props = I18nProps & {
-  error?: string,
+  autoFocus?: boolean,
+  error?: InputErrorMessage,
```
`React.ReactNode`

jacogr commented on packages/app-accounts/src/UploadButton.tsx:
```diff
@@ -0,0 +1,259 @@
+  address: string,
+  password: string,
+  isPasswordModalOpen: boolean,
+  error?: InputErrorMessage,
```
`React.ReactNode`

jacogr commented on packages/ui-app/src/Input.tsx:
```diff
@@ -2,20 +2,23 @@
   children?: React.ReactNode,
+  autoFocus?: boolean,
   defaultValue?: any,
+  error?: InputErrorMessage,
```
`React.ReactNode`

jacogr commented on packages/ui-app/src/InputError/index.tsx:
```diff
@@ -0,0 +1,38 @@
+import translate from '../translate';
+
+type Props = I18nProps & {
+  error: InputErrorMessage
```
`React.ReactNode`

jacogr commented on packages/ui-app/src/Notification.tsx:
```diff
@@ -0,0 +1,37 @@
+type SemanticMessageSizes = 'mini' | 'tiny' | 'small' | 'large' | 'big' | 'huge' | 'massive';
+
+type Props = BareProps & {
+  error?: InputErrorMessage,
```
`React.ReactNode`

jacogr commented on packages/ui-app/src/Password.tsx:
```diff
@@ -2,18 +2,22 @@
   children?: React.ReactNode,
   isDisabled?: boolean,
   isError?: boolean,
+  error?: InputErrorMessage,
```
`React.ReactNode`

jacogr commented on packages/ui-signer/src/Unlock.tsx:
```diff
@@ -14,14 +14,16 @@ import keyring from '@polkadot/ui-keyring/index';
 
 type State = {
   isError: boolean,
+  error?: InputErrorMessage,
```
`React.ReactNode`

jacogr commented on packages/app-accounts/src/DownloadButton.tsx:
```diff
@@ -0,0 +1,213 @@
+  address: string,
+  password: string,
+  isPasswordModalOpen: boolean,
+  error?: InputErrorMessage
```
`React.ReactNode` - this is the last one.

Repeating what I said in earlier comments - the CALLING component handles the translation, passing the string value (although it can now be a number, div, node, anything) to the underlying component.

ltfschoen commented:
![screen shot 2018-08-09 at 9 07 25 pm](https://user-images.githubusercontent.com/6226175/43920138-6191c206-9c18-11e8-93ee-19c655c716d0.png)

new look using the drag-and-drop!

ltfschoen commented:
@jacogr I've addressed all the comments, except for the following. I've written a response to clarify the requirements for each: 

* [ ] - Extra line before `store.set(accountKey(_address), json);`

**Response**: Would you like me to remove the extra line before `store.set(accountKey(_address), json);` in restore.ts?

* [ ] - No. className='accounts--flex-grow' (re `<div className={'accounts--flex-group-row'}>`), and for `<div className={'accounts--flex-item'}>`

**Response**: Sorry I don't quite understand. Are you suggesting I just rename that className to `accounts--flex-grow`, or would you like me to somehow incorporate `flex-grow` https://developer.mozilla.org/en-US/docs/Web/CSS/flex-grow?

* [ ] - Remove if we are not passing these through - as it should be React.ReactNode going in (re `export type UnlockI18n = {`)

**Response**: Existing code prior to this PR uses UnlockI18n in ui-signer and app-toolbox, I've left that existing implementation unchanged, but just moved their type definition into a shared file

* [ ] - Just isShort will do the trick here.
Is this not too small? Row or Summary maybe? (We can hide the balances/transactions if need-be))

**Response**: AddressMini is quite small, would be good for small screen sizes, i'll see how the other's look. I just tried implementing both AddressSummary and AddressMini in both DownloadButton.tsx and UploadButton.tsx. I was going to use media queries to toggle one or the other depending on the device width in use, or even try something like the react-responsive library.

I implemented it as follows:
```
import AddressMini from '@polkadot/ui-app/AddressMini';
import AddressSummary from '@polkadot/ui-app/AddressSummary';
...
              <div className={'accounts--Address-modal'}>
                <AddressMini isShort value={address} />
                <AddressSummary value={address} />
```

But when I added the AddressSummary component it gave error `Uncaught ExtError: Application root must be wrapped inside 'rx-react/Api' to provide API context`.

So I added the following, thinking it would resolve the error: 
```
import withApi from '@polkadot/ui-react-rx/with/api';
...
const Component: React.ComponentType<any> = translate(
  withApi(DownloadButton)
);
...
export default Component;
```

But it didn't resolve the error. What do you think?

jacogr commented:
1. Yes.
2. The class was initially named `flex-grow` - if changed to `accounts--flex-grow`. great
3. Ok, those need to be changed to React.ReactNode as well - for now probably best to leave as-is then.
4. Modals render inside a React Portal, so it doesn't pull through the nested areas of the main app. The Modal itself needs to pull API. (e.g. see ui-signer/Modal.tsx)

jacogr commented on packages/app-accounts/src/DownloadButton.tsx:
```diff
@@ -0,0 +1,204 @@
+        this.hidePasswordModal();
+      } else {
+        this.setState({
+          error: React.createElement('div', null, t('error', { defaultValue: 'Unable to obtain account from memory' }))
```
Do we need to create an element? Can't just pass the string? (React.ReactNode is basically any primitive, undefined or actual elements)

jacogr commented on packages/app-accounts/src/DownloadButton.tsx:
```diff
@@ -0,0 +1,204 @@
+      }
+    } catch (e) {
+      this.setState({
+        error: React.createElement('div', null, t('error', { defaultValue: 'Unable to save file' }))
```
As per above createElement comment.

jacogr commented on packages/app-accounts/src/DownloadButton.tsx:
```diff
@@ -0,0 +1,204 @@
+    }
+
+    return (
+      <div className={'accounts--Address-download'}>
```
Is the wrapping div needed? Cannot just pass the className to Modal?

jacogr commented on packages/app-accounts/src/DownloadButton.tsx:
```diff
@@ -0,0 +1,204 @@
+        >
+          <Modal.Content>
+            <div className='ui--grid'>
+              <div className={'accounts--Address-modal'}>
```
`<div className='accounts--Address-modal'>` (no extra wrapping `{}`'s

jacogr commented on packages/app-accounts/src/DownloadButton.tsx:
```diff
@@ -0,0 +1,204 @@
+          <Modal.Content>
+            <div className='ui--grid'>
+              <div className={'accounts--Address-modal'}>
+                <AddressMini isShort value={address} />
```
If an element has multiple attribute values, each appears on a seperate line.

jacogr commented on packages/app-accounts/src/Editor.spec.js:
```diff
@@ -0,0 +1,36 @@
+// of the ISC license. See the LICENSE file for details.
+
+import React from 'react';
+import { expect } from 'chai';
```
Why are we using chai? Everywhere else just use the jest built-in expect, e.g (picking a file) https://github.com/polkadot-js/common/blob/master/packages/util/src/hex/fixLength.spec.js#L9

jacogr commented on packages/app-accounts/package.json:
```diff
@@ -16,6 +16,12 @@
+    "@types/file-saver": "^1.3.0"
+  },
+  "devDependencies": {
+    "chai": "^4.1.2",
```
Remove chai.

jacogr commented on packages/app-accounts/src/Editor.tsx:
```diff
@@ -98,21 +100,39 @@ class Editor extends React.PureComponent<Props, State> {
+      return (
+        <div>
+          <div>{t('editor.nosaved', { defaultValue: 'There are no saved accounts. Create an account or upload a JSON file of a saved account.' })}</div>
+          <div className={'accounts--Address-wrapper'}>
```
`<div className='accounts--Address-wrapper'>` (no wrapping {})

jacogr commented on packages/app-accounts/src/Editor.tsx:
```diff
@@ -98,21 +100,39 @@ class Editor extends React.PureComponent<Props, State> {
+        <div>
+          <div>{t('editor.nosaved', { defaultValue: 'There are no saved accounts. Create an account or upload a JSON file of a saved account.' })}</div>
+          <div className={'accounts--Address-wrapper'}>
+            <div className={'accounts--Address-file'}>
```
`<div className='accounts--Address-file'>` (No wrapping {})

jacogr commented on packages/app-accounts/src/Editor.tsx:
```diff
@@ -98,21 +100,39 @@ class Editor extends React.PureComponent<Props, State> {
+          <div className={'accounts--Address-wrapper'}>
+            <div className={'accounts--Address-file'}>
+              <UploadButton
+                size='big'
```
Since this is only used in a single place - do we need a `size` prop? Set that as the default in the component.

jacogr commented on packages/app-accounts/src/Editor.tsx:
```diff
@@ -98,21 +100,39 @@ class Editor extends React.PureComponent<Props, State> {
-        <div className='grow'>
-          <div className='ui--row'>
+      <div className={'accounts--flex-group-row'}>
+        <div className={'accounts--flex-container-col-summary'}>
```
No need for wrapping {}'s

jacogr commented on packages/app-accounts/src/Editor.tsx:
```diff
@@ -98,21 +100,39 @@ class Editor extends React.PureComponent<Props, State> {
-          <div className='ui--row'>
+      <div className={'accounts--flex-group-row'}>
+        <div className={'accounts--flex-container-col-summary'}>
+          <div className={'accounts--flex-item'}>
```
No need for wrapping {}'s

jacogr commented on packages/app-accounts/src/Editor.tsx:
```diff
@@ -98,21 +100,39 @@ class Editor extends React.PureComponent<Props, State> {
+              value={address}
+            />
+          </div>
+          <div className={'accounts--flex-item-file'}>
```
No need for wrapping {}'s

jacogr commented on packages/app-accounts/src/Editor.tsx:
```diff
@@ -98,21 +100,39 @@ class Editor extends React.PureComponent<Props, State> {
+          </div>
+          <UploadButton onChangeAccount={this.onChangeAccount} />
+        </div>
+        <div className={'accounts--flex-container-col-inputs'}>
```
No need for wrapping {}'s

jacogr commented on packages/app-accounts/src/Editor.tsx:
```diff
@@ -98,21 +100,39 @@ class Editor extends React.PureComponent<Props, State> {
+          <UploadButton onChangeAccount={this.onChangeAccount} />
+        </div>
+        <div className={'accounts--flex-container-col-inputs'}>
+          <div className={'accounts--flex-item'}>
```
No need for wrapping {}'s

jacogr commented on packages/app-accounts/src/UploadButton.tsx:
```diff
@@ -0,0 +1,237 @@
+    }
+
+    const blob = new Blob([fileBytes], { type: 'text/plain;charset=utf-8' });
+
```
Extra blank line can be removed.

jacogr commented on packages/app-accounts/src/UploadButton.tsx:
```diff
@@ -0,0 +1,237 @@
+        if (!isUndefined(e) && e.target !== null) {
+          const fileContents: any = JSON.parse(e.target.result);
+
+          if (Object.keys(fileContents).includes('address' && 'encoding' && 'meta')) {
```
`'address' && 'encoding' && 'meta'` evaluates to -- `meta`

So the logic is wrong.

jacogr commented on packages/app-accounts/src/UploadButton.tsx:
```diff
@@ -0,0 +1,237 @@
+
+            // Store uploaded wallet in state and open modal to get their password for it
+            this.setState({ uploadedFileKeyringPair: json }, () => this.showPasswordModal());
+          }
```
What happens when the keys (see check above, it basically only checks for meta) does not find a valid file?

jacogr commented on packages/app-accounts/src/UploadButton.tsx:
```diff
@@ -0,0 +1,237 @@
+        onChangeAccount(pairRestored.publicKey());
+      } else {
+        this.setState({
+          error: React.createElement('div', null, t('error', { defaultValue: 'Unable to upload account into memory' }))
```
We don't need to create a div. See the previous comments.

jacogr commented on packages/app-accounts/src/UploadButton.tsx:
```diff
@@ -0,0 +1,237 @@
+    } catch (e) {
+      console.error('Error processing uploaded file to local storage: ', e);
+      this.setState({
+        error: React.createElement('div', null, t('error', { defaultValue: 'Unable to upload account into memory' }))
```
No need for React.createElement

jacogr commented on packages/app-accounts/src/UploadButton.tsx:
```diff
@@ -0,0 +1,237 @@
+    const address = uploadedFileKeyringPair && uploadedFileKeyringPair.address;
+
+    return (
+      <div className={'accounts--Address-upload'}>
```
No wrapping {}'s

jacogr replied:
Do we need to wrap the Modal? Can't just pass the className to the Modal?

jacogr commented on packages/app-accounts/src/UploadButton.tsx:
```diff
@@ -0,0 +1,237 @@
+        >
+          <Modal.Content>
+            <div className='ui--grid'>
+              <div className={'accounts--Address-modal'}>
```
No wrapping {}'s

jacogr commented on packages/app-accounts/src/UploadButton.tsx:
```diff
@@ -0,0 +1,237 @@
+          <Modal.Content>
+            <div className='ui--grid'>
+              <div className={'accounts--Address-modal'}>
+                <AddressMini isShort value={address} />
```
Multiple attrs each on their own line.

jacogr commented on packages/app-accounts/src/UploadButton.tsx:
```diff
@@ -0,0 +1,237 @@
+          </Modal.Content>
+        </Modal>
+        <File
+          className={'ui--Param-File-account'}
```
Unneeded {}'s

jacogr commented on packages/app-accounts/src/UploadButton.tsx:
```diff
@@ -0,0 +1,237 @@
+        <File
+          className={'ui--Param-File-account'}
+          withLabel
+          label={'upload account'}
```
Unneeded {}'s

jacogr commented on packages/app-accounts/src/UploadButton.tsx:
```diff
@@ -0,0 +1,237 @@
+          withLabel
+          label={'upload account'}
+          onChange={this.handleUploadedFiles}
+          acceptedFormats={'.json'}
```
Unneeded {}'s

jacogr commented on packages/app-accounts/src/UploadButton.tsx:
```diff
@@ -0,0 +1,237 @@
+          label={'upload account'}
+          onChange={this.handleUploadedFiles}
+          acceptedFormats={'.json'}
+          shouldDisplayFile={false}
```
If the prop is false, we don't need to pass it in. Remove.

jacogr commented on packages/app-accounts/test/enzyme.js:
```diff
@@ -0,0 +1,12 @@
+// This software may be modified and distributed under the terms
+// of the ISC license. See the LICENSE file for details.
+
+const Adapter = require('enzyme-adapter-react-16');
```
We may want to move this file to the root under test - it will be used by other apps as well. (No need to add duplication in that case)

jacogr commented on packages/ui-app/src/InputError/index.tsx:
```diff
@@ -0,0 +1,38 @@
+// // Copyright 2017-2018 @polkadot/ui-app authors & contributors
```
Remove.

jacogr commented on packages/ui-app/src/Notification.tsx:
```diff
@@ -0,0 +1,36 @@
+
+    return (
+      <div className='ui--Notifications'>
+        <Message size={size} error={!isUndefined(error)}>
```
Just `!!error` will do here.

jacogr replied:
Multiple props each on their own line.

jacogr commented on packages/ui-app/src/Password.tsx:
```diff
@@ -6,14 +6,18 @@ import { BareProps } from './types';
 
 import React from 'react';
 
+import classes from '@polkadot/ui-app/util/classes';
```
It is from the own package, remove `@polkadot/ui-app`, just `./util`

jacogr commented on packages/ui-keyring/src/account/load.ts:
```diff
@@ -0,0 +1,24 @@
+  const _address = json.address;
+
+  if (!_address) {
+    return;
```
Bubble error up and throw here, this is non-recoverable.

jacogr commented on packages/ui-keyring/src/account/backup.ts:
```diff
@@ -0,0 +1,38 @@
+  let jsonDecrypted = undefined;
+
+  if (!_address || !password) {
+    console.error('Missing address or password');
```
For all these cases it would probably make more sense to throw an exception instead of returning undefined - caller can handle it.

ltfschoen commented:
@jacogr I've addressed all the comments and replied some of them below:

* [x] - If an element has multiple attribute values, each appears on a seperate line. Multiple attrs each on their own line.
	* tslint props on multiple lines

  * **Response** - I tried modifying .eslintrc.json by adding the following "rules" property to try and detect this in future, but it doesn't appear to work (see https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-first-prop-new-line.md):

  ```
  {
	  "extends": [
	    "./node_modules/@polkadot/dev-react/config/eslint.js"
	  ],
	  "rules": {
	    "react/jsx-first-prop-new-lin": "multiline-multiprop"
	  }
	}
  ```

* [ ] - Since this is only used in a single place - do we need a size prop? Set that as the default in the component.

  * **Response** - I'm also using it in the same file when an address exists in the Editor view, but I don't want it to be big, as I want the icon to be the same size as the Copy (address) button. But I thought the UploadButton should be 'big' when no addresses exist, as it's the only icon shown in that situation

* [ ] - Do we need to wrap the Modal? Can't just pass the className to the Modal?

  * **Response** - Yes, I need a wrapping div in UploadButton since it's wrapping both the Modal and the File component. In DownloadButton we also need the wrapping div since it's wrapping the Modal and Button components.

* [ ] - Modals render inside a React Portal, so it doesn't pull through the nested areas of the main app. The Modal itself needs to pull API. (e.g. see ui-signer/Modal.tsx)

  * **Response** - I've tried to wrap the Modal with withApi in the following two branches with two separate attempts unsuccessfully in these commits:
    * https://github.com/polkadot-js/apps/commit/ff07e60c6792eee3155bf17e0637fa55236e24f8
    * https://github.com/polkadot-js/apps/commit/09d1896351709c90c3ac6a250e023af9092a6588

* [ ] - Unsure how to display full width password in the Modal for Upload and Download without passing the passwordWidth as props. I can't pass it with className because the className prop is used on a parent element instead of the one I want the styles applied to

jacogr commented:
- Ok, then we need at least one default - pick the place where it occurs most often, that is the default size, the other is the override.

- Modals rendering. Check, once again, ui-signer/Modal. The reason it does not work is related to the portal and having wrapping divs. The only way the injection works in this case is directly on the Modal. (as in signer, the Modal has the withApi)

- CSS had the ability to override, so for example, you can do `.accounts--uploadmodal .ui--Password .input { ... }`

ltfschoen commented:
@jacogr I've addressed the review requirements as follows:

1) I've removed the `size` prop, as it's not longer required since I'm using the drag-and-drop area now instead of an Upload icon

2) In the Upload and Download modal's it now shows the AddressSummary on large screensizes, and the AddressMini for small screensizes. I've added an extra optional prop to AddressMini to display the Copy icon next to it so it's consistent.

3) I've used CSS overrides as you suggested and it now the password extends across the full width of the modal, whilst in the Extrinsics modal it remains just half the width unchanged.

Also today you suggested moving the Download icon so that it appears on the right-hand side of the Copy icon (i.e. in the AddressSummary component), instead of where the Download icon is now in the Editor component just below the AddressSummary component. I've tried implementing that change but I can't get the tests to pass when I add that change, although it works and builds correctly. I've created a separate branch with a single commit that shows the changes I've made, https://github.com/polkadot-js/apps/commit/572d965bbe4c3fee4b21f0bb6955a0a2b8ee4e26. Initial after I made the changes it gave me errors in like `Error TS7023: 'render' implicitly has return type 'any' because it does not have a return type annotation and is referenced directly or indirectly in one of its return expressions`. I resolved them by implementing the recommendations at this link https://github.com/Microsoft/TypeScript/issues/8546 in AddressSummary.tsx by adding the following (notice the extra `: JSX.Element` and `<div></div>` instead of `null`, as recommended here: Microsoft/TypeScript#8546):
  ```
    protected renderAddress (): JSX.Element {
      ...
      {withDownloadButton ? <DownloadButton address={address} /> : <div></div>}
  ```
But after adding those changes I'm now getting the following errors, and haven't figured out how to resolve them:
```
Summary of all failing tests
 FAIL  packages/ui-app/build/AddressSummary.spec.js
  ● Test suite failed to run

    Configuration error:

    Could not locate module @polkadot/app-accounts/DownloadButton mapped as:
    /Users/scon/code/src/polkadot-js/apps/packages/ui-accounts/src//DownloadButton.

    Please check your configuration for these entries:
    {
      "moduleNameMapper": {
        "/@PolkaDot\/app-(accounts|addresses|example|explorer|extrinsics|rpc|staking|storage|toolbox|vanitygen)(.*)$/": "/Users/scon/code/src/polkadot-js/apps/packages/ui-$1/src/$2"
      },
      "resolver": null
    }

      at createNoMappedModuleFoundError (node_modules/jest-resolve/build/index.js:411:17)
      at Object.<anonymous> (packages/ui-app/build/AddressSummary.js:1438:48)

 FAIL  packages/ui-app/src/AddressSummary.spec.js
  ● Test suite failed to run

    Configuration error:

    Could not locate module @polkadot/app-accounts/DownloadButton mapped as:
    /Users/scon/code/src/polkadot-js/apps/packages/ui-accounts/src//DownloadButton.

    Please check your configuration for these entries:
    {
      "moduleNameMapper": {
        "/@PolkaDot\/app-(accounts|addresses|example|explorer|extrinsics|rpc|staking|storage|toolbox|vanitygen)(.*)$/": "/Users/scon/code/src/polkadot-js/apps/packages/ui-$1/src/$2"
      },
      "resolver": null
    }

      at createNoMappedModuleFoundError (node_modules/jest-resolve/build/index.js:411:17)
      at Object.<anonymous> (packages/ui-app/src/AddressSummary.tsx:935:42)

 FAIL  packages/app-accounts/src/DownloadModal.spec.js
  ● Test suite failed to run

    Configuration error:

    Could not locate module @polkadot/app-accounts/DownloadButton mapped as:
    /Users/scon/code/src/polkadot-js/apps/packages/ui-accounts/src//DownloadButton.

    Please check your configuration for these entries:
    {
      "moduleNameMapper": {
        "/@PolkaDot\/app-(accounts|addresses|example|explorer|extrinsics|rpc|staking|storage|toolbox|vanitygen)(.*)$/": "/Users/scon/code/src/polkadot-js/apps/packages/ui-$1/src/$2"
      },
      "resolver": null
    }

      at createNoMappedModuleFoundError (node_modules/jest-resolve/build/index.js:411:17)
      at Object.<anonymous> (packages/ui-app/src/AddressSummary.tsx:935:42)

 FAIL  packages/app-accounts/build/DownloadModal.spec.js
  ● Test suite failed to run

    Configuration error:

    Could not locate module @polkadot/app-accounts/DownloadButton mapped as:
    /Users/scon/code/src/polkadot-js/apps/packages/ui-accounts/src//DownloadButton.

    Please check your configuration for these entries:
    {
      "moduleNameMapper": {
        "/@PolkaDot\/app-(accounts|addresses|example|explorer|extrinsics|rpc|staking|storage|toolbox|vanitygen)(.*)$/": "/Users/scon/code/src/polkadot-js/apps/packages/ui-$1/src/$2"
      },
      "resolver": null
    }

      at createNoMappedModuleFoundError (node_modules/jest-resolve/build/index.js:411:17)
      at Object.<anonymous> (packages/ui-app/src/AddressSummary.tsx:935:42)

 FAIL  packages/app-accounts/src/Editor.spec.js
  ● Test suite failed to run

    Configuration error:

    Could not locate module @polkadot/app-accounts/DownloadButton mapped as:
    /Users/scon/code/src/polkadot-js/apps/packages/ui-accounts/src//DownloadButton.

    Please check your configuration for these entries:
    {
      "moduleNameMapper": {
        "/@PolkaDot\/app-(accounts|addresses|example|explorer|extrinsics|rpc|staking|storage|toolbox|vanitygen)(.*)$/": "/Users/scon/code/src/polkadot-js/apps/packages/ui-$1/src/$2"
      },
      "resolver": null
    }

      at createNoMappedModuleFoundError (node_modules/jest-resolve/build/index.js:411:17)
      at Object.<anonymous> (packages/ui-app/src/AddressSummary.tsx:935:42)

 FAIL  packages/app-accounts/build/Editor.spec.js
  ● Test suite failed to run

    Configuration error:

    Could not locate module @polkadot/app-accounts/DownloadButton mapped as:
    /Users/scon/code/src/polkadot-js/apps/packages/ui-accounts/src//DownloadButton.

    Please check your configuration for these entries:
    {
      "moduleNameMapper": {
        "/@PolkaDot\/app-(accounts|addresses|example|explorer|extrinsics|rpc|staking|storage|toolbox|vanitygen)(.*)$/": "/Users/scon/code/src/polkadot-js/apps/packages/ui-$1/src/$2"
      },
      "resolver": null
    }
```

ltfschoen commented:
@jacogr I've incorporated your review comments and passed down DownloadButton as buttonChildren prop to AddressSummary and it worked great, thanks!

jacogr commented on packages/app-accounts/src/DownloadButton.tsx:
```diff
@@ -0,0 +1,148 @@
+        this.hidePasswordModal();
+      } else {
+        this.setState({
+          error: 'Unable to obtain account from memory'
```
Translatable strings, we never us strings without `t(<key>, { defaultValue: <value> })`. So here it would be -

```js
handleDownloadAccount = (): void => {
  const { t } = this.props;

...

this.setState({
  error: t('download.error.memory', {
    defaultValue: 'Unable to obtain account from memory'
});
...
}
```

jacogr commented on packages/app-accounts/src/DownloadButton.tsx:
```diff
@@ -0,0 +1,148 @@
+      }
+    } catch (e) {
+      this.setState({
+        error: 'Unable to save file'
```
Same as above. i18n strings. Always.

jacogr commented on packages/app-accounts/src/DownloadButton.tsx:
```diff
@@ -0,0 +1,148 @@
+          handleDownloadAccount={this.handleDownloadAccount}
+          hidePasswordModal={this.hidePasswordModal}
+          isPasswordModalOpen={isPasswordModalOpen}
+          key='accounts-download-signer-modal'
```
Any reason for the key?

jacogr commented on packages/app-accounts/src/DownloadButton.tsx:
```diff
@@ -0,0 +1,148 @@
+    const { address } = this.props;
+
+    return {
+      address: address,
```
Just 

```js
return {
  address,
  passord: '',
...
```

jacogr commented on packages/app-accounts/src/DownloadModal.tsx:
```diff
@@ -0,0 +1,129 @@
+              <div className='accounts--Address-modal-message expanded'>
+                <p>
+                  <Trans i18nKey='unlock.info'>
+                    Please enter your account password to unlock and download a decrypted backup.
```
Why not only `t(...)`

jacogr commented on packages/app-accounts/src/DownloadModal.tsx:
```diff
@@ -0,0 +1,129 @@
+        onChange={onChangePassword}
+        password={password}
+        value={keyringAddress.publicKey()}
+        tabIndex='1'
```
Number.

jacogr commented on packages/app-accounts/src/DownloadModal.tsx:
```diff
@@ -0,0 +1,129 @@
+          <Button
+            isNegative
+            onClick={onDiscard}
+            tabIndex='3'
```
Number.

jacogr commented on packages/app-accounts/src/DownloadModal.tsx:
```diff
@@ -0,0 +1,129 @@
+            isPrimary
+            className='ui--Button-submit'
+            onClick={handleDownloadAccount}
+            tabIndex='2'
```
Number.

jacogr commented on packages/app-accounts/src/Editor.tsx:
```diff
@@ -15,6 +15,8 @@ import accountObservable from '@polkadot/ui-keyring/observable/accounts';
 import withObservableBase from '@polkadot/ui-react-rx/with/observableBase';
 
 import AddressSummary from '@polkadot/ui-app/AddressSummary';
+import UploadButton from './UploadButton';
```
Blank line before. Remove blank line before preceding import (which is in the same library block)

jacogr commented on packages/app-accounts/src/Editor.tsx:
```diff
@@ -88,23 +90,37 @@ class Editor extends React.PureComponent<Props, State> {
+        <div className='accounts--flex-container-col-summary'>
+          <div className='accounts--flex-item'>
+            <AddressSummary
+              buttonChildren={<DownloadButton address={address || ''} />}
```
No need for the `|| ''` - handle this in the underlying component.

jacogr commented on packages/app-accounts/src/Editor.tsx:
```diff
@@ -88,23 +90,37 @@ class Editor extends React.PureComponent<Props, State> {
+            <AddressSummary
+              buttonChildren={<DownloadButton address={address || ''} />}
+              className='shrink'
+              value={address || ''}
```
As above, empty should be handled by the component being wrapped.

jacogr commented on packages/app-accounts/src/UploadButton.tsx:
```diff
@@ -0,0 +1,180 @@
+          // Cast to type 'any' since property 'result' does exist on type 'EventTarget'
+          // Reference: https://stackoverflow.com/a/45017155/3208553
+          const fileContents: any = JSON.parse((e.target as any).result);
+          const expectedJsonProperties = ['address', 'encoding', 'meta'];
```
Hoist these constants to file-level. (i.e. it does not need to be created as a new variable each time the function is executed)

jacogr commented on packages/app-accounts/src/UploadButton.tsx:
```diff
@@ -0,0 +1,180 @@
+      } catch (e) {
+        console.error('Error uploading file: ', e);
+        this.setState({
+          error: 'Unable to upload account from file'
```
Translatable strings.

jacogr commented on packages/app-accounts/src/UploadButton.tsx:
```diff
@@ -0,0 +1,180 @@
+        onChangeAccount(pairRestored.publicKey());
+      } else {
+        this.setState({
+          error: 'Unable to upload account into memory'
```
Translatable strings. As per earlier comments.

jacogr commented on packages/app-accounts/src/UploadButton.tsx:
```diff
@@ -0,0 +1,180 @@
+    } catch (e) {
+      console.error('Error processing uploaded file to local storage: ', e);
+      this.setState({
+        error: 'Unable to upload account into memory'
```
Translatable strings.

jacogr commented on packages/app-accounts/src/UploadButton.tsx:
```diff
@@ -0,0 +1,180 @@
+          error={error}
+          hidePasswordModal={this.hidePasswordModal}
+          isPasswordModalOpen={isPasswordModalOpen}
+          key='accounts-upload-signer-modal'
```
Unneeded key?

jacogr commented on packages/app-accounts/src/UploadButton.tsx:
```diff
@@ -0,0 +1,180 @@
+        <File
+          className='ui--Param-File-account'
+          withLabel
+          label='upload account'
```
Translatable strings.

jacogr commented on packages/app-accounts/src/UploadModal.tsx:
```diff
@@ -0,0 +1,117 @@
+              <div className='accounts--Address-modal-message expanded'>
+                <p>
+                  <Trans i18nKey='unlock.info'>
+                    Please enter your account password to upload and restore it encrypted.
```
This could just be `t(...)`

jacogr commented on packages/app-accounts/src/UploadModal.tsx:
```diff
@@ -0,0 +1,117 @@
+          <Button
+            isNegative
+            onClick={onDiscard}
+            tabIndex='3'
```
Number.

jacogr commented on packages/app-accounts/src/UploadModal.tsx:
```diff
@@ -0,0 +1,117 @@
+          <Button
+            isPrimary
+            onClick={processUploadedFileStorage}
+            tabIndex='2'
```
Number.

jacogr commented on packages/app-accounts/src/UploadModal.tsx:
```diff
@@ -0,0 +1,117 @@
+        error={error}
+        onChange={onChangePassword}
+        password={password}
+        tabIndex='1'
```
Number, i.e. `={1}`

jacogr commented on packages/ui-app/src/Button/types.d.ts:
```diff
@@ -18,6 +18,7 @@ export type ButtonProps = BareProps & {
   isPrimary?: boolean,
   onClick?: () => void | Promise<void>,
   size?: Button$Sizes,
+  tabIndex?: string | number,
```
`number` only.

jacogr commented on packages/ui-app/src/Input.tsx:
```diff
@@ -28,6 +31,7 @@ type Props = BareProps & {
   name?: string,
   onChange: (value: string) => void,
   placeholder?: string,
+  tabIndex?: number | string,
```
`number`

jacogr commented on packages/ui-app/src/Params/Param/File.tsx:
```diff
@@ -19,7 +19,9 @@ type Props = BareProps & {
   t: TranslationFunction,
-  withLabel?: boolean
+  withLabel?: boolean,
+  acceptedFormats?: string,
```
Alphabetical ordering.

jacogr commented on packages/ui-app/src/Params/Param/File.tsx:
```diff
@@ -39,7 +41,7 @@ class BytesFile extends React.PureComponent<Props, State> {
 
   render () {
-    const { className, isDisabled, isError = false, label, placeholder, t, withLabel } = this.props;
+    const { className, isDisabled, isError = false, label, placeholder, t, withLabel, acceptedFormats, shouldDisplayFile } = this.props;
```
Alphabetical ordering.

jacogr commented on packages/ui-app/src/Params/Param/File.tsx:
```diff
@@ -52,11 +54,12 @@ class BytesFile extends React.PureComponent<Props, State> {
           className={classes('ui--Param-File', isError ? 'error' : '', className)}
           disabled={isDisabled}
           multiple={false}
+          accept={acceptedFormats}
```
Order.

jacogr commented on packages/ui-app/src/Password.tsx:
```diff
@@ -6,17 +6,22 @@ import { BareProps } from './types';
   label?: string,
   name?: string,
   onChange: (value: string) => void,
+  tabIndex?: number | string,
```
`number` only.

jacogr commented on packages/ui-app/src/Password.tsx:
```diff
@@ -61,6 +69,7 @@ export default class Password extends React.PureComponent<Props, State> {
           }
           isPrimary
           onClick={this.onToggleVisible}
+          tabIndex='-1'
```
`={-1}` (preferably leave it out)

jacogr commented on packages/ui-app/src/util/arrayContainsArray.spec.js:
```diff
@@ -0,0 +1,68 @@
+
+    expect(actualOutput).toEqual(expectedOutput);
+  });
+  it('will return false if superset has all items in the subset but one is misspelt', () => {
```
Extra blank line before block.

jacogr commented on packages/ui-signer/src/Unlock.tsx:
```diff
@@ -13,14 +13,17 @@ import keyring from '@polkadot/ui-keyring/index';
+  error?: React.ReactNode,
   onChange: (password: string) => void,
   password: string,
+  tabIndex?: number | string,
```
`number`

ltfschoen commented:
Progress update:

* Fixed all bugs in commit 8d386383806094c5abbd8a6ed99ed58f31352279

**Remaining issues**

* Move Restore Account to separate tab (i.e. [Edit Account] (or) [ Create Account ] (or) [Restore Account])

* Add more unit tests

jacogr commented:
If our components only allow `number`, it is quite valid to pass that number through to a underlying component that accepts `number | string`. So that should not be an issue at all.

ltfschoen commented:
@jacogr I've made updates that integrate with the latest master including Tabs component, and I've replaced ` tabIndex?: number | string` with `tabIndex?: number`. 

**Remaining tasks**

Add unit tests using ReactTestUtils in the interim due to difficulties with Enzyme's `setState`. Another developer has now reported that they're encountering the same problem https://github.com/airbnb/enzyme/issues/1794

jacogr commented on packages/app-accounts/src/DownloadButton.tsx:
```diff
@@ -0,0 +1,163 @@
+
+  static getDerivedStateFromProps ({ address }: Props, prevState: State): State {
+    return {
+      address: address
```
Just

```
return {
  address
};
```

jacogr commented on packages/app-accounts/src/util/accounts.ts:
```diff
@@ -0,0 +1,27 @@
+let showIsOrAre: (accountAll?: any[] | undefined) => string;
+let showPlural: (accountAll?: any[] | undefined) => string;
+
+accountsQty = (accountAll?: any[] | undefined): number => isAccounts(accountAll) && accountAll ? Object.keys(accountAll).length : 0;
```
Using `accountsAll?: any[]` implies `undefined` as well, i.e. the `?` does that, it says "this can be supplied or not"

jacogr commented on packages/app-accounts/src/util/accounts.ts:
```diff
@@ -0,0 +1,27 @@
+
+accountsQty = (accountAll?: any[] | undefined): number => isAccounts(accountAll) && accountAll ? Object.keys(accountAll).length : 0;
+
+isAccounts = (accountAll?: any[] | undefined): boolean => !isNoAccounts(accountAll);
```
Same as above. (And for the rest of the file)

jacogr commented on packages/app-accounts/src/Editor.tsx:
```diff
@@ -85,150 +75,54 @@ class Editor extends React.PureComponent<Props, State> {
                 defaultValue: 'using my account'
               })}
-              onChange={this.onChangeAccount}
+              onChange={onChangeAccount}
```
I am not convinced about moving any of this logic to index.tsx - the base component does not care about editing, it is only concerned with rendering the correct Components.

jacogr commented on packages/app-accounts/src/index.tsx:
```diff
@@ -68,6 +118,129 @@ class AccountsApp extends React.PureComponent<Props, State> {
     this.setState({ action: 'edit' });
   }
+
+  selectRestore = (): void => {
```
As per the comment about moving the rest of the functionality from the Editor above. What we have now is one monster component that contains all the logic for the whole app, instead of the different pieces in the seperate components where they belong. 

Basically we are creating an un maintainable monster here.

ltfschoen replied:
Ok, I just felt I needed `accountAll` in index.tsx so I'd know whether any accounts existed in order to decide whether to render the Creator or Editor component, since we want it to automatically switch from Editor view to Creator view if the user clicks to forget all their accounts. So I'm deciding between whether to add the `accountAll` observable into index.tsx or to store a boolean state variable `isAccounts` in index.tsx that is updated by child components.

If the user restores an account in the Restorer component view I thought I'd need to set the restored keypair as the `current` keypair in the parent state of index.tsx so that the router would know to automatically switch to the Editor component view, with the `current` keypair shown in the selection box

jacogr commented on packages/app-accounts/src/index.tsx:
```diff
@@ -68,6 +118,129 @@ class AccountsApp extends React.PureComponent<Props, State> {
+    );
+  }
+
+  nextState (newState: State = {} as State): void {
```
All these new added functionality, revert to what it was so we have self-contained pieces and not a god component.

jacogr commented on packages/app-accounts/src/util/inputSubmitWithEnter.ts:
```diff
@@ -0,0 +1,17 @@
+// of the ISC license. See the LICENSE file for details.
+
+// Allow user to submit form by pressing enter whilst input field still focussed
+export default function inputSubmitWithEnter (event: any, node: any): void {
```
Surely this is not specific to accounts? This is Input-level stuff.

ltfschoen commented:
@jacogr I've pushed some changes:

* Pushed this commit https://github.com/polkadot-js/apps/pull/202/commits/22d61032b2140899b3283fdca308c6a2f8e9106a that:
	* Replaced `accountAll?: any[] | undefined` with `accountAll?: any[]`
	* Simply returned address with `{ address }`
	* Editing logic of `onChangeAccount` moved back into Editor. Only calls parent index routing component to switch to rendering correct component (like OnBack)
	* Moved `inputSubmitWithEnter` into ui-app
	* Reverted god/monster component back to what it was previously with self-contained components 
		* Except added observable of accountAll in index, to know when to update the tabs UI (i.e. to hide Editor menu tab when no accounts exist, but to show it again when an account is Restored)


* Pushed alternative solution in commit https://github.com/polkadot-js/apps/commit/097ae2819c9fc65d30e4ee173a326c4212c92fa8 of separate branch luke-74-download-upload-with-password-no-observe-accountall
	* Tries to avoid having to add accountAll observable in index, but doesn't look nice at all...

jacogr commented:
The pulling in of accountsAll is fine, it makes sense since it makes the UI dynamic. The idea is good. (Basically with no accounts, if you go to `#accounts` there is really no reason to have the `Edit` visible, it is not applicable at all - and the message is helpful, but as it stands requires an additional click)

jacogr commented on packages/app-accounts/src/Creator.tsx:
```diff
@@ -23,7 +23,7 @@ import AddressSummary from '@polkadot/ui-app/AddressSummary';
 
 type Props = I18nProps & {
-  onBack: () => void
+  onCreateAccount: () => void
```
Great naming here. (Much better than original version)

jacogr commented on packages/app-accounts/src/Creator.tsx:
```diff
@@ -207,15 +207,25 @@ class Creator extends React.PureComponent<Props, State> {
+    // add encrypted data to keyring since immediately after account creation only secret key is in memory
+    keyring.loadAccount(json);
+
+    this.onCreate();
```
We can actually just call directly to `onCreateAccount` here instead of having the member function that add nothing special, it just calls the function, nothing else.

jacogr commented on packages/app-accounts/src/Restorer.tsx:
```diff
@@ -0,0 +1,56 @@
+      <div>
+        <div className='accounts--Restorer-message'>
+          {t('restorer.existing', {
+            defaultValue: `There ${showIsOrAre(accountAll)} ${accountsQty(accountAll)} saved account${showPlural(accountAll)}. Create an account or upload a JSON file of a saved account.`
```
Instead of using string concat, which changes with values, use templates.

```
t('restorer.existing', {
  defaultValue: 'There {{quantity}} saved accounts. Create an account or upload a JSON file of a saved account.',
  replace: {
    quantity: accountsQty(accountAll)
  }
})
```

i18next support is/are & plurals as well, so no need to bring in those functions. (Which in itself is specific to the English language and therefore non-translatable)

ltfschoen replied:
@jacogr I read through the following links that explain `replace`, `count`, and how to use plurals using the `_plural` postfix:
* https://www.i18next.com/translation-function/essentials
* https://www.i18next.com/translation-function/plurals

I then tried adding the following code, but regardless of what value is assigned to the  `count` property (which the docs say is the value used for plurals), it always just displays the value assigned to `defaultValue` in the UI. i.e. if `count` is 0, instead of returning the value assigned to `defaultValue_0` that is prefixed with `There are ` it instead just returns the `defaultValue` prefixed with `There is `, which isn't grammatically correct, and it does the same of just displaying the value assigned to `defaultValue` even if there are multiple accounts assigned to the `count` property, instead of returning the value assigned to `defaultValue_plural`.
```
{t('restorer.existing', {
  defaultValue: 'There is {{count}} saved account. {{instruction}}',
  defaultValue_0: 'There are {{count}} saved accounts. {{instruction}}',
  defaultValue_plural: 'There are {{count}} saved accounts. {{instruction}}',
  replace: {
    instruction: 'Create an account or upload a JSON file of a saved account.',
    count: accountsQty(accountAll)
  }
})}
```

jacogr commented on packages/app-accounts/src/UploadButton.tsx:
```diff
@@ -0,0 +1,205 @@
+import UploadModal from './UploadModal';
+import translate from './translate';
+
+let actualJsonProperties: Array<string> = [];
```
Why are there not in the class? They are supposed to be singletons? (i.e. exist and have value across instances?)

jacogr commented on packages/app-accounts/src/UploadButton.tsx:
```diff
@@ -0,0 +1,205 @@
+let fileContents: Object = {};
+const fileReader: FileReader = new FileReader();
+let json;
+let pairRestored;
```
All class-based apart from the function.

jacogr commented on packages/app-accounts/src/UploadButton.tsx:
```diff
@@ -0,0 +1,205 @@
+
+let actualJsonProperties: Array<string> = [];
+let blob: Blob;
+const createBlob = (fileBytes: Uint8Array): Blob => new Blob([fileBytes], { type: 'text/plain;charset=utf-8' });
```
Readability, prefer -

```
const createBlob = (fileBytes: Uint8Array): Blob =>
  new Blob([fileBytes], { type: 'text/plain;charset=utf-8' });
```

jacogr commented on packages/app-accounts/src/util/accounts.ts:
```diff
@@ -0,0 +1,27 @@
+
+isNoAccounts = (accountAll?: Array<any>): boolean => !accountAll || !Object.keys(accountAll).length;
+
+showIsOrAre = (accountAll?: Array<any>): string => accountsQty(accountAll) === 0 || accountsQty(accountAll) > 1 ? 'are' : 'is';
```
We cannot do this. See the comment earlier about plurals.

jacogr commented on packages/app-accounts/src/util/accounts.ts:
```diff
@@ -0,0 +1,27 @@
+// This software may be modified and distributed under the terms
+// of the ISC license. See the LICENSE file for details.
+
+let accountsQty: (accountAll?: Array<any>) => number;
```
why not -

```
const accountsQt = (accountAll?: Array<any>): number =>
  isAccounts(accountAll) && accountAll ? Object.keys(accountAll).length : 0;
```

ltfschoen commented:
@jacogr I've addressed the review comments and replied about difficulties showing singular/plural. I've also made changes to render nothing until we know if any accounts exist or not when a user goes to http://localhost:3000/#/accounts because the default action is still Edit, but when it was loading it was showing the Create and Restore tabs but without either of the tabs being "active" yet, and it was showing account 5xx.xxx.

jacogr commented:
It looks good and does what it say on the tin. The heading looks very long, I played a bit and simply made it "Confirm account removal" (Just as a test)

<img width="556" alt="polkadot apps portal 2018-09-07 10-14-05" src="https://user-images.githubusercontent.com/1424473/45207053-1e17b780-b287-11e8-9507-47ba85c2298c.png">

If we want to have something more descriptive, we can say exactly what it does under the account info with just normal-weight and size text, i.e. tell the user that this removes the account from the browser, etc. (Optional, but it provides more guidance which is what we orginally had with the much longer title)

jacogr commented:
Ahah! That is where that comment went. Dangit, I knew I was not crazy.

ltfschoen commented:
@jacogr I've merged the latest from master, renamed `accountall` to `allAccounts` so it's consistent with the Transfer PR, I've reverted to using a generic message in Restore. I've fixed a bug so the Download button appears in the Editor view next to the copy button again. I removed the observable of `allAccounts` from Editor since it wasn't being used in that component. And lastly, I've tried to add an Enzyme/Jest test for the Editor component that sets the state so an account exists and then tries to check whether the DownloadButton component is rendered, however I've been difficulty getting the test to pass. When I check if the DownloadButton length is 1, it says it expected 1 but actually returned the following:
```
{Symbol(enzyme.__root__): {Symbol(enzyme.__root__): [Circular], Symbol(enzyme.__unrendered__): <Editor t={[Function mockT]} />, Symbol(enzyme.__renderer__): {"batchedUpdates": [Function batchedUpdates], "getNode": [Function getNode], "render": [Function render], "simulateError": [Function simulateError], "simulateEvent": [Function simulateEvent], "unmount": [Function unmount]}, Symbol(enzyme.__node__): {"instance": null, "key": undefined, "nodeType": "host", "props": {"children": [<Translate(Forgetting) … />, <div … />, <ButtonGroup … />], "className": "accounts--Editor"}, "ref": null, "rendered": [{"instance": null, "key": undefined, "nodeType": "class", "props": [Object], "ref": null, "rendered": null, "type": [Function Translate]}, {"instance": null, "key": undefined, "nodeType": "host", "props": [Object], "ref": null, "rendered": [Array], "type": "div"}, {"instance": null, "key": undefined, "nodeType": "class", "props": [Object], "ref": null, "rendered": [Array], "type": [Function ButtonGroup]}], "type": "div"}, Symbol(enzyme.__nodes__): [{"instance": null, "key": undefined, "nodeType": "host", "props": {"children": [Array], "className": "accounts--Editor"}, "ref": null, "rendered": [[Object], [Object], [Object]], "type": "div"}], Symbol(enzyme.__options__): {"adapter": {"options": {"enableComponentDidUpdateOnSetState": true, "lifecycles": [Object]}}, "attachTo": undefined, "hydrateIn": undefined}}, Symbol(enzyme.__unrendered__): null, Symbol(enzyme.__renderer__): {"batchedUpdates": [Function batchedUpdates], "getNode": [Function getNode], "render": [Function render], "simulateError": [Function simulateError], "simulateEvent": [Function simulateEvent], "unmount": [Function unmount]}, Symbol(enzyme.__node__): undefined, Symbol(enzyme.__nodes__): [], Symbol(enzyme.__options__): {"adapter": {"options": {"enableComponentDidUpdateOnSetState": true, "lifecycles": {"componentDidUpdate": [Object], "getDerivedStateFromProps": true, "getSnapshotBeforeUpdate": true, "setState": [Object]}}}, "attachTo": undefined, "hydrateIn": undefined}}
```

If I use `mount` instead of `shallow`, I get the following error when running the tests ` Error: Uncaught [TypeError: Cannot read property 'options' of undefined]`, which I overcame by mocking I18n with:
``` 
jest.mock('react-i18next', () => ({
  // this mock makes sure any components using the translate HoC receive the t function as a prop
  translate: () => Component => {
    Component.defaultProps = { ...Component.defaultProps, t: k => k };
    return Component;
  }
}));
```
But when I run the test after doing that I get the following error `Error: Uncaught [ExtError: Application root must be wrapped inside 'rx-react/Api' to provide API context]`, but I'm not sure how to mock it. So far I've tried the following without success:
```
jest.mock('@polkadot/ui-react-rx/Api');
```

jacogr commented:
If you want to use mount, you will need to create the Api context and wrap the component in it. Or maybe even wrap it in `createApp` from `ui-app`?

ltfschoen commented:
@jacogr I've added a test to check that the address of the current pair is passed down as a prop to the DownloadButton component of AddressSummary

ltfschoen commented on packages/app-accounts/src/UploadModal.tsx:
```diff
@@ -0,0 +1,120 @@
+
+    return (
+      <Modal
+        className={classes('ui--accounts-upload-Signer', className)}
```
I'm assuming that we're using BEM http://getbem.com/naming/. 
I used the `-Signer` postfix in UploadModal so it would be consistent with the naming convention used in ui-signer's Modal. I'll remove the `ui--`prefix as you've suggested. I'm going to propose to make it `accounts--UploadModal-Signer` (since that'd be consistent with how it's done everywhere else). I'll make wrapper div in UploadButton `accounts--UploadButton`

ltfschoen commented:
@jacogr I've pushed commits that address all the latest review comments, with the exception of one of them:

> packages/ui-app/src/Params/Param/File.tsx
> 
> I'm trying to wrap my head around this - Params are for Parameters as received from Extrinsics, Storage and/or submitted for those. So how and why does our file from accounts tie into this? If we are re-using, there should really rather be a component extracted from this (into ui-app/InputFile) that can be used by this and others.


 I'll continue going through this review comment tomorrow morning when my internet connection isn't playing up.

ltfschoen commented:
@jacogr I've finished pushing commits that address all the review comments, including creation and use of InputFile

ltfschoen commented:
@jacogr Oh, I just found a bug whereby the filename is not displaying after uploading a file in both Extrinsics and Accounts, I'll look into that now

ltfschoen commented:
@jacogr I've fixed the bug so it shows the filename being uploaded by default

ltfschoen commented:
@jacogr I just merged latest from master and discovered some of the styles weren't correct anymore, so I've fixed them.

jacogr commented on packages/ui-app/src/Notification.tsx:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    return (
+      <div className='ui--Notifications'>
```
Since it is `ui-app` Component, we need to take care of

- passing-in styles (overrides)
- passing in className (append via classes)

ltfschoen commented:
@jacogr I've addressed the review comments including the nice-to-have

ltfschoen commented:
@jacogr I've addressed your latest review comments in the latest four commits and written detailed commit messages explaining the changes

ltfschoen commented:
Closing in favour of using PR #333 instead. Possible additional changes to PR #333 have been proposed to be merged from PR #336.
